Project 2 description

brson opened the pull request:
Fixes https://github.com/pingcap/talent-plan/issues/11

brson commented:
After reviewing bitcask I went back through this and adjusted various text to reflect how bitcask works. I added a (unfinished) section on "database architecture" that is supposed to be an architectural overview. It also defines a consistent terminology to use for the various components throughout the course. Please give them some thought.

sticnarf commented on rust/projects/file-io/project.md:
```diff
@@ -64,17 +94,22 @@ methods:
-state cached in memory.
+sequential log. On startup, the commands in the log are re-evaluated and the log
+pointer (file offset) of the last command to modify each key recorded in the
+in-memory index.
```
I have difficulty in understanding this sentence. I find the sentence structure weird, and I am confused by "the last command" here.

sticnarf replied:
Ah, I now understand it. We find the last command which modifies each key, and record the log pointer of it in the in-memory index.

sticnarf replied:
Can you try to make the sentence more understandable?

brson replied:
Done.

sticnarf commented on rust/projects/file-io/project.md:
```diff
@@ -145,24 +180,33 @@ disk.
 
 - "set"
   - The user invokes `kvs set mykey, myvalue`
   - `kvs` creates a struct representing the "get" command, containing its key and
```
```suggestion
   - `kvs` creates a struct representing the "set" command, containing its key and
```

sticnarf commented on rust/projects/file-io/project.md:
```diff
@@ -145,24 +180,33 @@ disk.
+  - It then appends the serialized command to a file containing the log
   - If that succeeds, it exits silently with error code 0
   - If it fails, it exits by printing the error and return a non-zero error code
 - "get"
```
Do we really need to state how "get" works here in "Part 2: Storing writes in the log"?

sticnarf replied:
Shall we move the more general behavior of the two subcommands to "Project spec" and leave the details in these two parts?

brson replied:
Perhaps not. I was thinking that describing "get" would give context to the "set" algorithm, and since it's described here it is _not_ described in the section implementing get.

brson replied:
For the moment I've left it per my explanation. Feel free to object.

sticnarf commented on rust/projects/file-io/project.md:
```diff
@@ -187,22 +236,164 @@ to read about the "get" command. It may help to keep both in mind, or to
+_Implement "get" now_.
+
+
+# Part 5: Storing log pointers in the index
```
Part 4 instead? The following parts should be modified too.

sticnarf replied:
And should use h2 instead of h1. Same for part 5 and 6.

brson replied:
Fixed. Good catch.

sticnarf commented on rust/projects/file-io/project.md:
```diff
@@ -187,22 +236,164 @@ to read about the "get" command. It may help to keep both in mind, or to
+changes to disk, then stays resident in memory to service future
+queries (it is stateful).
+
+Is your `KvsStore` stateful or stateless?
```
```suggestion
Is your `KvStore` stateful or stateless?
```

sticnarf commented on rust/projects/file-io/project.md:
```diff
@@ -187,22 +236,164 @@ to read about the "get" command. It may help to keep both in mind, or to
+the log? That is, how do you know when to stop reading one record, and start
+reading the next? Do you even need to? Maybe serde will deserialize a record
+directly from an I/O stream and stop reading when it's done, leaving the
+[cursor] in the correct place to read subsequent records. Maybe serde will
```
What will "cursor" link to?

brson replied:
"file cursor" on wikipedia. But instead I'll just say "file cursor" and our audience will probably know what that is.

brson replied:
Fixed.

sticnarf commented on rust/projects/file-io/project.md:
```diff
@@ -187,22 +236,164 @@ to read about the "get" command. It may help to keep both in mind, or to
+
+Remember that "get" may not find a value and that case has to be handled
+specially. Here, our API returns `None` and our command line client prints
+a particular message and exits with a non-zero exit code.
```
zero or non-zero? Conflicts with the above described behavior.

brson replied:
zero. will change

sticnarf commented on rust/projects/file-io/project.md:
```diff
@@ -187,22 +236,164 @@ to read about the "get" command. It may help to keep both in mind, or to
+
+_Implement log compaction for your database._
+
+With compaction, what is the amortized complexity (big O) of calling `set`? What
```
I believe the complexity is O(n^2) using a fixed threshold (such as 1MB mentioned above). Are we misleading on purpose?

sticnarf commented on rust/projects/file-io/project.md:
```diff
@@ -64,17 +94,22 @@ methods:
+until it is dropped.
 
 
 ## Project setup
```
Ask students to add necessary dev-dependencies so that tests can run.

brson replied:
Done.

brson commented:
> I think the number of data files does not make much difference, but a separate index file counts. If we don't have a index file, we need to read (or seek through) the full data file (even after compaction) on rebuilding our index table. Loading a small index file is way faster instead.

This lesson should create the index file, yes. For now I've added it as a todo.

With a single log essentially the entire database needs to be scanned every compaction if all the stale entries are to be found. With multiple files, I gather bitcask uses, it's simple to compact fractions of the database at once.

> I believe the complexity is O(n^2) using a fixed threshold (such as 1MB mentioned above). Are we misleading on purpose?

In the section you are referring to the thing I was trying to hint at is that every once in a while "set"
triggers a massive compaction, while most "sets" only do work proportional to the size of the key & value. So there's a massive discrepancy in latency between the average and worst cases.

What is "n" in your analysis of n^2?

I think in the obvious solution, where every once in a while the entire log needs to be scanned and copied, the complexity is dominated by the size of the database. So I'm thinking the complexity is linear in the size of the database, `O(d)`. In practice though, there isn't a compaction on every set, so the work is amortized, which may have a different complexity class.

I haven't thought about this very hard, but it's obvious that if there's a compaction every byte written, then every byte costs the size of the database, `d`. With a compaction every two bytes, then every byte costs `d/2`. Every three, `d/3`. So maybe the amortized complexity is `O(d/b)` where b is the average number of bytes between compactions.

Because that might be off base I've put todos there.

And because I think the "total on-disk size of all un-compacted log entries" is poorly defined, and maybe a bad choice I've put a big todo on the whole compaction description.

Also, I removed one of the three tables to make them less boring.

Thanks for the great review @sticnarf .

sticnarf commented:
@brson 
> With a single log essentially the entire database needs to be scanned every compaction if all the stale entries are to be found. With multiple files, I gather bitcask uses, it's simple to compact fractions of the database at once.

I don't think I fully understand what you mean. With multiple files, the improvement I can see is to record where the stale entries locate and just merge those files with stale entries. Do you mean that? I think it's generally unavoidable to scan the whole database in a compaction, in case there are stale entries in every old files.

> What is "n" in your analysis of n^2?
>
> I haven't thought about this very hard, but it's obvious that if there's a compaction every byte written, then every byte costs the size of the database, d. With a compaction every two bytes, then every byte costs d/2. Every three, d/3. So maybe the amortized complexity is O(d/b) where b is the average number of bytes between compactions. Edit: that is probably also just O(d).

`n` is the number of bytes written. I didn't know how to measure the amortized complexity of writing in a database. To perform an amortized analysis, I believed it was inappropriate to use a changing variable "d" (size of the database). After a second look, I find that, under certain assumptions (such as 50% writes create stale data while the others not), the complexity can just not converge. Here I post my analysis if you are interested:
>
> Say we will write a total of `m` times with `n` bytes every time, and `n` is far less than the threshold `T` (maybe 1M). To be more concrete, let's say, if 1MB is written, 0.5MB is new and create 0.5MB stale data. And, if stale data reach 1MB, we do a compaction.
>
> We will trigger about `(mn)/(2T)` times of compaction in total. Each compaction needs to write bytes linear to database size. Then, the total bytes written in the compaction "W" is:
>
> W = T + 2T + 3T + ... + mn/2 = (T + mn/2) * (mn)/(2T) / 2 = mn(T + mn / 2) / (4T)
>
> The amotized bytes written each time is W / m = n(T + mn/2) / (4T) = O(mn^2). When m approaches infinity, the amotized bytes written also approach infinity.

So now I think such an assumption does not make sense in reality. Maybe using database size is appropriate. 

Of course, we can have students think of an algorithm heuristic enough.

> And because I think the "total on-disk size of all un-compacted log entries" is poorly defined, and maybe a bad choice I've put a big todo on the whole compaction description.

Yes, I had the feeling that it was poorly defined, too.

sticnarf commented:
LGTM. Thank you for great text. @brson

brson commented:
> I don't think I fully understand what you mean. With multiple files, the improvement I can see is to record where the stale entries locate and just merge those files with stale entries. Do you mean that? I think it's generally unavoidable to scan the whole database in a compaction, in case there are stale entries in every old files.

Yes, that is what I mean. There are two ways I can think of to improve average request latency (not necessarily complexity) by having multiple logs:

- Compaction can be done incrementally, even as a background thread. This would mean that a request never has to wait for a full compaction.
- Track which segments of the full log are dirty and only compact those.

A practical implementation can't block requests on full compactions, or maybe any compactions. It seems very likely to me that moving from full log compactions to compacting only dirty segments of the log would change the complexity analysis significantly.

At some point I need to understand more precisely how bitcask treats the log.





> We will trigger about (mn)/(2T) times of compaction in total. Each compaction needs to write bytes linear to database size. Then, the total bytes written in the compaction "W" is:
> 
> > W = T + 2T + 3T + ... + mn/2 = (T + mn/2) * (mn)/(2T) / 2 = mn(T + mn / 2) / (4T)
> 
> The amotized bytes written each time is W / m = n(T + mn/2) / (4T) = O(mn^2). When m approaches infinity, the amotized bytes written also approach infinity.

Thanks for the analysis! It's better than I did for sure. I understand the approach you are taking but don't follow it to the end. I am bad at math though. I have to move on now, but we can consider it more later.

This is a pretty important question for an implementer to consider, and not trivial per this thread, so I'm inclined to keep it in some form. For now I've left todos with links back to this discussion.
